Add option to not advertise software codecs as supported

luca-schlecker opened the pull request:
Sometimes, software-codecs can cause performance-related issues like seen in #4838. The added option allows disabling software codecs, forcing a transcode to a non-software (hardware-accelerated) codec.

The latest version did not fix the issue described in #4838 for me. The software decoder still struggled with my particular file, even though the resolution was within the advertised range.

**Changes**
- A new checkbox in the _Advanced Playback Preferences_, toggling whether or not software codecs are used.
- An accompanying new user preference (`use_software_codecs`)
- Updated logic in `createDeviceProfile` and `MediaCodecCapabilitiesTest` to allow excluding software-based codecs.
- A new translatable string `pref_use_software_codecs`.

**Code assistance**
No AI-based assistance tools were used.

**Issues**
I don't know of an open issue for this.

luca-schlecker committed:
Add option to not advertise software codecs as supported

Sometimes, software-codecs can cause performance-related issues like
seen in #4838. The added option allows disabling software codecs,
forcing a transcode to a non-software/hardware-accelerated codec.

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/util/profile/MediaCodecCapabilitiesTest.kt:
```diff
@@ -308,4 +315,16 @@ class MediaCodecCapabilitiesTest(
+
+			// detect software-based codecs by name; approach taken from here:
+			// https://android.googlesource.com/platform/frameworks/av/+/d5dbc2b31c492be23fc4d4a96619330b63499b06/media/libstagefright/MediaCodecList.cpp#364
+			name.startsWith("OMX.google.", ignoreCase = true) || name.startsWith("c2.android.", ignoreCase = true)
```
The codecs available on a device are highly vendor specific. This line assumes a lot of things and will definitely be wrong on a lot of devices. We should ONLY trust the `isSoftwareOnly` property, which means this check can't work on Android 9 and older.

luca-schlecker replied:
Ok, I've removed the problematic check and instead always return false there. This results in the default behavior prior to this change if `isSoftwareOnly` is not supported. Should there be an extra description in the settings option to communicate this limitation to the user?

nielsvanvelzen replied:
The setting should be hidden on those devices as it won't function.

luca-schlecker replied:
The settings entry is now hidden if it is not supported by the device. 👍

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/preference/UserPreferences.kt:
```diff
@@ -259,6 +259,11 @@ class UserPreferences(context: Context) : SharedPreferenceStore(
+		/**
+		 * Enable the use of software-based codecs.
+		 */
+		var useSoftwareCodecs = booleanPreference("use_software_codecs", true)
```
```suggestion
		var softwareCodecsEnabled = booleanPreference("software_codecs_enabled", true)
```

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/util/profile/MediaCodecCapabilitiesTest.kt:
```diff
@@ -308,4 +315,12 @@ class MediaCodecCapabilitiesTest(
 		return Size(maxWidth, maxHeight)
 	}
+
+	private fun MediaCodecInfo.isSoftwareCodec(): Boolean {
```
function could be simplified to a getter

```kt
private val MediaCodecInfo.isSoftwareCodec: Boolean
	get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q && isSoftwareOnly
```

nielsvanvelzen commented on app/src/main/res/values-de/strings.xml:
```diff
@@ -646,4 +646,5 @@
     <string name="playback_buffer_extra_large">Sehr groß</string>
     <string name="pref_burn_subtitles_when_transcoding">Untertitel beim Transkodieren einbrennen</string>
     <string name="pref_burn_subtitles_when_transcoding_description">Untertitel beim Transkodieren immer einbrennen. Dies kann Fehler bei der Untertitel-Synchronisation beheben, aber verhindert individualisierte Untertitel.</string>
+    <string name="pref_use_software_codecs">Software-Codecs verwenden</string>
```
Please do not push translations via pull requests. Only the `values/strings.xml` should be changed, other files need to be managed via translate.jellyfin.org after merging.

nielsvanvelzen commented on app/src/main/res/values-en-rGB/strings.xml:
```diff
@@ -642,4 +642,5 @@
     <string name="sdr">SDR</string>
     <string name="hdr10">HDR10</string>
     <string name="hlg">HLG</string>
+    <string name="pref_use_software_codecs">Use software codecs</string>
```
Please do not push translations via pull requests. Only the `values/strings.xml` should be changed, other files need to be managed via translate.jellyfin.org after merging.

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/util/profile/deviceProfile.kt:
```diff
@@ -98,7 +99,10 @@ fun createDeviceProfile(
 	userHEVCLevel: Int?,
+	useSoftwareCodecs: Boolean,
 ) = buildDeviceProfile {
+	mediaTest.useSoftwareCodecs = useSoftwareCodecs
```
No static properties, pass variable via called functions

luca-schlecker replied:
I've opted for making it a parameter of the primary constructor of the `MediaCodecCapabilitiesTest` class as it is used in multiple functions across it. 👍

luca-schlecker committed:
Add option to not advertise software codecs as supported

Sometimes, software-codecs can cause performance-related issues like
seen in jellyfin#4838. The added option allows disabling software
codecs, forcing a transcode to a non-software/hardware-accelerated
codec.

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/settings/screen/playback/SettingsPlaybackAdvancedScreen.kt:
```diff
@@ -216,6 +217,18 @@ fun SettingsPlaybackAdvancedScreen() {
+					headingContent = { Text(stringResource(R.string.pref_software_codecs_enabled)) },
+					trailingContent = { Checkbox(checked = softwareCodecsEnabled) },
+					onClick = { softwareCodecsEnabled = !softwareCodecsEnabled }
+				)
```
Should add a caption here with some more info about what this option does, I'd say something like how it is more likely to transcode with this disabled.

luca-schlecker replied:
How do you like _"Some software-based implementations may degrade performance. Disabling this option makes transcoding more likely."_ as a description?

nielsvanvelzen replied:
Sounds good, just need to make sure it fits in the UI but I think it will.

luca-schlecker replied:
Will test it this evening.

nielsvanvelzen replied:
Yeah seems to fit fine

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/settings/screen/playback/SettingsPlaybackAdvancedScreen.kt:
```diff
@@ -216,6 +217,18 @@ fun SettingsPlaybackAdvancedScreen() {
 		}
 
+		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+			item {
```
I'd say it makes more sense to put this setting on the "media codecs" page

luca-schlecker replied:
Will move it there 👍
